### DISCUSSION TITLE
NextJS not working with React sortable tree

### DISCUSSION THREAD

**Discussion by JithinKS97** [vercel/next.js#13553] [Help] [answered]
This is the react-sortable-tree repo
https://github.com/frontend-collective/react-sortable-tree

The error that I get in the browser when I installed react sortable tree in next project is this

Server Error
Error: Must use import to load ES Module: /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/dist/esm/index.js
require() of ES modules is not supported.

require() of /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/dist/esm/index.js from /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/frontend-collective-react-dnd-scrollzone/lib/index.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/dist/esm/index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/package.json.

BTW I use typescript in next

Please help me to resolve this

**Comment by jamesmosier** [comment, marked as answer]
You'll need to compile this module in order for it to work. ES modules do not get compiled by Next currently.

There are a few ways to do this, one is to [use this next-transpile-modules](https://github.com/martpie/next-transpile-modules) plugin for your `next.config.js`

**Comment by JithinKS97** [comment]
Thanks so much for the very quick reply. I installed next-transpile-modules

included this in next.config.js
```
const withTM = require("next-transpile-modules")([
  "react-dnd",
  "react-sortable-tree",
]);
module.exports = withTM();
```
But still, I get the error

Error: Must use import to load ES Module: /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/dist/esm/index.js
require() of ES modules is not supported.
require() of /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/dist/esm/index.js from /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/frontend-collective-react-dnd-scrollzone/lib/index.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/dist/esm/index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /home/jithinks/Desktop/rework/dynamiclearning/client/node_modules/react-dnd/package.json.

I also tried including   "frontend-collective-react-dnd-scrollzone", but no luck so far

Thanks so much

**Reply by JithinKS97** [reply]
Do I need to make some changes in babel config?

**Reply by jamesmosier** [reply]
Could you provide a code sample?

**Comment by JithinKS97** [comment]
@jamesmosier I've created a repo and reproduced the error
https://github.com/JithinKS97/next-react-sortable-tree-error
I've been able to solve the issue by downgrading the react-sortable-tree version from 2.7.1 to 2.6.2

**Reply by jamesmosier** [reply]
Thanks for the example repo! It appears the issue is that `react-sortable-tree`, although exports their package as a cjs (commonjs) package, are not bundling their dependencies as cjs imports (they are still ESM which won't work due to lack of support in Node, where Next runs your build). So all of the imports they perform are not transpiled as they need to be.

**Reply by JithinKS97** [reply]
OK. What can we do about this? Also the same package (React sortable tree) worked fine in Meteor. Does that mean meteor is doing some intermediate transpiling?

**Comment by wilgnne** [comment]
I was able to resolve the error ```Must use import to load ES Module``` using typescript
```js
{
  "compilerOptions": {
    "lib": [
      "dom",
      "dom.iterable",
      "esnext",
      "es2018"
    ],
  }
}
```
but I run into a new error, ```ReferenceError: window is not defined```

**Comment by softmarshmallow** [comment]
Solution -> https://github.com/react-dnd/react-dnd/issues/3416#issuecomment-1126159287
